File: Day 2/cSharp/day2-part2.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

#if PART2
class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Day 2 Part 2 started!");

        try
        {
            string input = File.ReadAllText("input-day2.md").Trim();
            string[] ranges = input.Split(',');
            var invalidIds = new List<long>();

            foreach (var range in ranges)
            {
                string[]  rangeParts = range.Split('-');
                long start = long.Parse(rangeParts[0]);
                long end = long.Parse(rangeParts[1]);
                for (long i = start; i <= end; i++)
                {
                    string numStr = i.ToString();
                    var options = GetEqualDivisionOptions(numStr);
                    foreach (var option in options)
                    {
                        var parts = DivideIntoEqualParts(numStr, option);
                        if (AllPartsAreTheSameTheyreRottingMyBrain(parts))
                        {
                            invalidIds.Add(i);
                            break;
                        }
                    }
                }
            }
            Console.WriteLine(invalidIds.Sum());
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error reading or processing input: {ex.Message}");
        }
    }

    static List<int> GetEqualDivisionOptions(string num)
    {
        var options = new List<int>();
        for (int i = 2; i <= num.Length; i++)
        {
            if (num.Length % i == 0)
            {
                options.Add(i);
            }
        }
        return options;
    }

    static List<string> DivideIntoEqualParts(string numStr, int numParts)
    {
        var parts = new List<string>();
        int partLength = numStr.Length / numParts;
        for (int i = 0; i < numParts; i++)
        {
            parts.Add(numStr.Substring(i * partLength, partLength));
        }
        return parts;
    }


    //:3 :3 :3 :3 :3 :3
    static bool AllPartsAreTheSameTheyreRottingMyBrain(List<string> parts)
    {
        return parts.All(part => part == parts[0]);
    }
}
#endif

File: Day 2/day2-part2.js
const fetchInput = require("./fetchInput");

console.log("Day 2 Part 2 started!");

function getAllInvalidIds(input) {
  const invalidIds = [];
  const ranges = input.trim().split(",");
  for (const range of ranges) {
    const [id1, id2] = range.split("-");
    let num1 = Number(id1);
    let num2 = Number(id2);
    while (num1 <= num2) {
      const numStr = num1.toString();
      const options = getEqualDivisionOptions(numStr);
      for (const option of options) {
        const parts = divideIntoEqualParts(numStr, option);
        if (allPartsAreTheSameTheyreRottingMyBrain(parts)) {
          invalidIds.push(num1);
          break;
        }
      }
      num1++;
    }
  }
  return invalidIds;
}

function getEqualDivisionOptions(num) {
  const options = [];
  for (let i = 2; i <= num.length; i++) {
    if (num.length % i === 0) {
      options.push(i);
    }
  }
  return options;
}

function divideIntoEqualParts(numStr, numParts) {
  const parts = [];
  const partLength = numStr.length / numParts;
  for (let i = 0; i < numParts; i++) {
    parts.push(numStr.slice(i * partLength, (i + 1) * partLength));
  }
  return parts;
}

//:3 :3 :3 :3 :3 :3
function allPartsAreTheSameTheyreRottingMyBrain(parts) {
  return parts.every((part) => part === parts[0]);
}

function sumInvalidIds(invalidIds) {
  let total = 0;
  for (let i = 0; i < invalidIds.length; i++) {
    total += invalidIds[i];
  }
  return total;
}

async function main() {
  try {
    const rawInput = await fetchInput();

    console.log(sumInvalidIds(getAllInvalidIds(rawInput)));
  } catch (error) {
    console.error("Error fetching or processing input:", error);
  }
}

main();
